xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesService.java

```diff
@@ -96,43 +102,57 @@ public void createServiceEndpointWatcher(CreateServiceEndpointWatcherRequest req
                 "Create watcher: " + watcherName, watcher, author, true));
     }
 
-    private static void validateWatcherAndPush(
+    private void validateWatcherAndPush(
             StreamObserver<ServiceEndpointWatcher> responseObserver,
             ServiceEndpointWatcher watcher, Runnable onSuccess) {
         // Create a KubernetesEndpointGroup to check if the watcher is valid.
         // We use KubernetesEndpointGroup for simplicity, but we will implement a custom implementation
         // for better debugging and error handling in the future.
-        final KubernetesEndpointGroup kubernetesEndpointGroup = createKubernetesEndpointGroup(watcher);
-
-        final AtomicBoolean completed = new AtomicBoolean();
-        final CompletableFuture<List<Endpoint>> whenReady = kubernetesEndpointGroup.whenReady();
-        final ServiceRequestContext ctx = ServiceRequestContext.current();
-
-        // Use a schedule to time out the watcher creation until we implement a custom implementation.
-        final ScheduledFuture<?> scheduledFuture = ctx.eventLoop().schedule(() -> {
-            if (!completed.compareAndSet(false, true)) {
-                return;
-            }
-            kubernetesEndpointGroup.closeAsync();
-            responseObserver.onError(
-                    Status.INTERNAL.withDescription(
-                            "Failed to retrieve k8s endpoints within 5 seconds. watcherName: " +
-                            watcher.getName()).asRuntimeException());
-        }, 5, TimeUnit.SECONDS);
-
-        whenReady.handle((endpoints, cause) -> {
-            if (!completed.compareAndSet(false, true)) {
-                return null;
-            }
-            scheduledFuture.cancel(false);
-            kubernetesEndpointGroup.closeAsync();
+        final ContextAwareBlockingTaskExecutor taskExecutor =
+                ServiceRequestContext.current().blockingTaskExecutor();
+        final CompletableFuture<KubernetesEndpointGroup> future =
+                createKubernetesEndpointGroup(watcher, xdsResourceManager.xdsProject().metaRepo(),
+                                              taskExecutor);
+        future.handle((kubernetesEndpointGroup, cause) -> {
```

Comment:
Should we safely call `future.join()` because the services are marked as `@Blocking`?

Reply:
Yeah, we can do that, but I wanted to avoid blocking the thread as much as possible.
Please let me know if you still think we should still use `future.join`.